### PR TITLE
Move spin_loop_hint to core::hint module

### PR DESCRIPTION
**Pull request by clarfonthey**
As mentioned in #55002. The new name is kept unstable to decide whether the function should have `_hint` in its name.

### CODE REVIEW

**Comment by KodrAus** [comment]
_sad trombone_

@bors retry

**Comment by clarcharr** [comment]
I honestly don't know how to fix the error here. I was under the impression that you could set the stability of an export differently from the original, but that might have changed.

**Comment by KodrAus** [comment]
It's a bit unfortunate.

How about we punt on the feature gate, decide in #55002 whether the name should be `core::hint::spin_loop` or `core::hint::spin_loop_hint` (throwing my hat in with `core::hint::spin_loop`), and go from there?

File: src/libcore/sync/atomic.rs [hunk 1]
```diff
@@ -84,28 +84,8 @@ use intrinsics;
...
-    unsafe {
-        asm!("pause" ::: "memory" : "volatile");
-    }
-
-    #[cfg(target_arch = "aarch64")]
-    unsafe {
-        asm!("yield" ::: "memory" : "volatile");
-    }
-}
+pub use hint::spin_loop as spin_loop_hint;
```
Comment (KodrAus) [inline src/libcore/sync/atomic.rs]: Any reason not to deprecate this one?

Comment (clarcharr) [reply src/libcore/sync/atomic.rs]: Because if the other name is unstable, then you can't actually use the replacement.

Comment (KodrAus) [reply src/libcore/sync/atomic.rs]: Ah of course :) I'd already put myself in the mindset that this would be stable.

**Comment by clarcharr** [comment]
Do you think that it would be worthwhile simply doing an FCP here to name the replacement `spin_loop`? Not really sure how we can initiate conversation on this short of starting an FCP.

I can update the PR to make the renamed version instantly stable and deprecate the old one, if that's desired.

**Comment by KodrAus** [comment]
> Do you think that it would be worthwhile simply doing an FCP here to name the replacement spin_loop?

Sure, let's do that :+1: @rfcbot doesn't actually like me though so I'll just give the libs team a ping.

cc @rust-lang/libs what do you think about deprecating `core::sync::atomic::spin_loop_hint` in favour of `core::hint::spin_loop`?

If we don't want the insta-stable `core::hint::spin_loop` function then we might just have to do some more shimming shenanigans to make the `core::sync::atomic::spin_loop_hint` function internally call `core::hint::spin_loop`.

**Comment by clarcharr** [comment]
That's a good idea-- I'll update the PR to do that regardless and we can decide whether to insta-stabilise or not.

**Comment by KodrAus** [comment]
@bors try

Ok, based on @alexcrichton's comments I think we can move ahead with the unstable `renamed_spin_loop` feature and do an FCP in the tracking issue to stabilize it, then deprecate the other later.

How does that sound?

**Comment by clarcharr** [comment]
Sounds good to me!

**Comment by kennytm** [comment]
(Cleaning up the bors status)

**Comment by clarcharr** [comment]
Fixed broken link; should be ready to merge now.

**Comment by KodrAus** [comment]
@bors r+

Thanks @clarcharr! I'll update the tracking issue.

**Pull request merged**